File: haskell/src/Y2015/D06.hs
{-# LANGUAGE RecordWildCards, LambdaCase, TupleSections, FlexibleContexts #-}

module Y2015.D06
  ( d06
  ) where

import LibPuzzle

import Data.Array
import Text.Parsec

d06 :: Computation
d06 = do
  instructions <- fmap parseInstruction . lines <$> readPuzzle 2015 6 "input"
  let grid       = mkGrid instructions
      lit        = length . filter fst . elems $ grid
      brightness = sum . fmap snd . elems $ grid
  answers
    [ answer lit        $ show lit ++ " light(s) is/are lit"
    , answer brightness $ show brightness ++ ": total brightness"
    ]

data Action = Off | Toggle | On

data Instruction = Instruction
  { action         :: Action
  , li, lj, ui, uj :: Int
  }

parseInstruction :: String -> Instruction
parseInstruction = fromRight' . parse instruction ""
  where fromRight' = either (error . show) id
        instruction =
          action >>= \act ->
          coords >>= \(li, lj) -> through >>
          coords >>= \(ui, uj) -> eof >>
          pure (Instruction act li lj ui uj)
          where trim f  = f >>= \a -> spaces >> pure a
                through = trim $ string "through"
                action  = trim $ on <|> off <|> toggle
                  where on     = try (string "turn on")  >> pure On
                        off    = try (string "turn off") >> pure Off
                        toggle = try (string "toggle")   >> pure Toggle
                coords  = trim $
                  num >>= \i -> char ',' >>
                  num >>= \j -> pure (read i, read j)
                  where num = many1 digit

mkGrid :: [Instruction] -> Array (Int, Int) (Bool, Int)
mkGrid instructions = accumArray update (False, 0) ((0, 0), (999, 999)) actionList
  where update (state, brightness) = \case
          On     -> (True, brightness + 1)
          Off    -> (False, max 0 (brightness - 1))
          Toggle -> (not state, brightness + 2)
        actionList = [((i, j), action) | Instruction {..} <- instructions, i <- [li..ui] , j <- [lj..uj]]

File: rust/src/y2015/d06.rs
use self::Action::{Toggle, TurnOff, TurnOn};
use crate::{parse::*, Date, Day, Puzzle, Result};
use std::{
    cmp::{max, min},
    str::FromStr,
};

const DATE: Date = Date::new(Day::D06, super::YEAR);
pub(super) const PUZZLE: Puzzle = Puzzle::new(DATE, solve);

const SIZE: usize = 1000;

#[allow(clippy::needless_pass_by_value)]
fn solve(input: String) -> Result {
    let instructions = parse_instructions(&input)?;
    let mut grid = Grid::default();
    for instruction in instructions {
        grid.apply(instruction);
    }
    let lit = grid.lit.iter().filter(|&&lit| lit).count();
    let brightness = grid.brightness.iter().sum::<u32>();
    answer!(lit, brightness);
}

fn parse_instructions(s: &str) -> Result<Vec<Instruction>> {
    s.lines().map(|line| line.parse()).collect()
}

enum Action {
    TurnOff,
    Toggle,
    TurnOn,
}

struct Coord {
    row: usize,
    col: usize,
}

struct Instruction {
    action: Action,
    start: Coord,
    end: Coord,
}

impl FromStr for Instruction {
    type Err = failure::Error;

    #[rustfmt::skip]
    fn from_str(s: &str) -> Result<Self> {
        named!(action(Str<'_>) -> Action, alt!(
            tag!("turn off") => { |_| TurnOff } |
            tag!("toggle")   => { |_| Toggle  } |
            tag!("turn on")  => { |_| TurnOn  }
        ));
        named!(coord(Str<'_>) -> Coord, do_parse!(
            col: uint!(usize) >> tag!(",") >>
            row: uint!(usize) >>
            (Coord { row, col })
        ));
        nom!(
            sep!(Str(s), space0, do_parse!(
                action: action >>
                start: coord >> tag!("through") >>
                end: coord >>
                ({
                    let Coord { row: row1, col: col1 } = start;
                    let Coord { row: row2, col: col2 } = end;
                    let start = Coord {
                        row: min(row1, row2),
                        col: min(col1, col2),
                    };
                    let end = Coord {
                        row: max(row1, row2),
                        col: max(col1, col2),
                    };
                    Self { action, start, end }
                })
            ))
        )
    }
}

struct Grid {
    lit: Box<[bool]>,
    brightness: Box<[u32]>,
}

impl Default for Grid {
    fn default() -> Self {
        Self {
            lit: vec![false; SIZE * SIZE].into_boxed_slice(),
            brightness: vec![0; SIZE * SIZE].into_boxed_slice(),
        }
    }
}

impl Grid {
    fn apply(&mut self, instruction: Instruction) {
        let start = instruction.start;
        let end = instruction.end;
        for row in start.row..=end.row {
            let i = row * SIZE + start.col;
            let j = row * SIZE + end.col;
            for lit in &mut self.lit[i..=j] {
                match instruction.action {
                    TurnOff => *lit = false,
                    Toggle => *lit = !*lit,
                    TurnOn => *lit = true,
                }
            }
            for brightness in &mut self.brightness[i..=j] {
                match instruction.action {
                    TurnOff => *brightness = brightness.saturating_sub(1),
                    Toggle => *brightness += 2,
                    TurnOn => *brightness += 1,
                }
            }
        }
    }
}
